### DISCUSSION TITLE
Opening the content black box

### DISCUSSION THREAD

**Discussion by memeplex** [typst/typst#4745] [Q&A] [answered]
Hi,

I'm implementing a template for my slideshows, in part to keep it simple and focused to my specific needs, in part to improve my typst coding skills. Although I dislike when things get too magical, I decided to map level 1 headings to separator slides and level 2 headings to regular slides (usually my go-to would have been to implement functions in order to capture the semantics of each kind of slide and to have finer and simpler control over the structure, but I wanted a challenge).

Another requirement is that the body of regular slides must be centered both horizontally and vertically. I mean, the body considered as a single block (despite https://github.com/typst/typst/issues/3605, that can be worked around anyway). But for that I need to separate the heading from the rest of the page:

```
heading

align(..., body)
```

Here is where things get tricky. My first attempt was to set a show rule for pages and then process their content. My hope was to find a heading neatly followed by paragraphs and stuff, but that's not quite what I found. First, the (func of the body of the) page is a flow, which is an internal structure AFAICS (there is something about it in this [Laurenz' post](https://laurmaedje.github.io/posts/layout-models/)). Nevertheless, the flow has a children field with an array of its content. But this content has already gone through a number of transformations, e.g. compare:

```typst
#show page: it => {
  let c = it.body.children
  if (c.len() > 2) { c } else { it }
}

= Heading

Par1

Par2

- Item1
- Item2
```

<img width="427" alt="image" src="https://github.com/user-attachments/assets/b062c097-2874-4027-9adf-ec71a3d189f7">

to

```typst
#let f(c) = c.children

#f([
= Heading

Par1

Par2

- Item1
- Item2
])
```

<img width="427" alt="image" src="https://github.com/user-attachments/assets/9bfb88d7-2f37-4890-8358-9e094af4cf22">

For some things there is more structure/semantics, e.g. already packaged lists instead of loose items, paragraphs instead of text separated by parbreaks. For others is the other way around: sadly, the heading isn't there anymore. I could split the first 3 children of the flow, and manipulate the rest, but it would be as bad and fragile as hacks get.

So my next attempt was to put a label or some metadata in a heading show rule, signaling the end of the heading and the beginning of the body. Unfortunately, neither labels nor metadata will do till the page rule, they don't appear in the flow at all.

Another attempt was to take advantage of `query`, but there seems to be no way to get raw content between two locations (namely, two second level headings): selectors match specific functions, there is no "wildcard selector".

So at this point I'm short of ideas and a bit confused about those transformations of content and would like to ask:

1. Is mixed content (as opposed to content produced by a single function in [Model](https://typst.app/docs/reference/model/)) supposed to be treated as a black box that can be concatenated, wrapped, etc. but never opened? If that is the case, then all those transformations aren't my business.

2. Then again, if that is the case, what would be the "typst philosophy" here? Write your own function that clearly separates blocks with different semantics? This would be a fine answer to me and, as I said, it's closer to the way I would have done it, but here I'm trying to understand and to push the envelope a bit. Also, this subsection = slide thing seems to be the touchstone of slideshow packages for some reason. And it plays better with the outline produced by my LSP. But those are moot points.

3. Is there another obvious way to achieve what I want that I'm, perhaps shamefully, missing?

Thanks!

**Comment by laurmaedje** [comment, marked as answer]
> Is mixed content (as opposed to content produced by a single function in [Model](https://typst.app/docs/reference/model/)) supposed to be treated as a black box that can be concatenated, wrapped, etc. but never opened?

Mixed content is indeed supposed to be treated as a black box. Once `context` comes in, you cannot peek into it anymore anyway. That you can observe the internal flow at all is not ideal and doing that will most likely not be possible anymore in the next version.

Ultimately content is just a bunch of values of different types[^1] nested in each other to represent some structure that is to be transformed into something visible by Typst. At the shallowest level, looking at content fields is useful: The fields of an element in a show rule, for instance, allow you to redefine how that element should look based on its part. You _can_ transform more of the structure yourself, but it's often not a good idea because it is complex to account for all cases. As a rule of thumb, I would say it is probably a bad idea unless you know that the content will have a specific fixed structure.

> Then again, if that is the case, what would be the "typst philosophy" here? Write your own function that clearly separates blocks with different semantics?

Typically yes. I can imagine more powerful show rules that allow selection of a sequence of matching elements (sort of like regex) in the future and those would unlock a few more use cases and would allow writing something behaving like `list` (being collected from `list.item`) in "user space".

[^1]: In the future, the plan is to turn each element function into a different type and a content element just becomes a normal value.

**Reply by memeplex** [reply]
Thanks for the clarifications! As a Typst noob I was misled by two red herrings I guess:

1. That show rules look similar to CSS so you want to do everything mainly that way, perhaps with a bit of "js" here and there.

2. The fact that you can peep into the contents of content gives you the feeling that they are something akin to Pandoc AST, a predictable structure that is there to be manipulated.

As per your answer, I can forget about 2. Regarding 1 I know there are plans and proposals to make the mechanism more powerful ([inside/outside selectors](https://github.com/typst/typst/issues/1926), [custom types with set/show rules](https://laurmaedje.github.io/posts/types-and-context/), [constraining match counts](https://github.com/typst/typst/issues/1926#issuecomment-2268086789), sequences of matching elements -as you have just mentioned-).

But both CSS and Pandoc markdown ultimately require the annotation of the document at the inline/span level and block/div level. There you can only hope to react, whereas Typst functions can directly act on the pieces and also instrument them with local set/show rules, alleviating the need for more complex selectors or AST transformers.

IMO set/show rules are mostly important in order to deal with lightweight markup syntax that doesn't provide the knobs and pulls of function calling and also to fine-tune the building blocks internally used by the more complex builtin elements. But simple packages providing a template and maybe a few functions, and that cannot extend the syntax, can go a long way without them. There is some minor inconvenience in providing functions tied by common settings that you'd rather not pass each time (as set rules facilitate), but that can be easily surmounted by creating closures from an initialization function (i.e. closures as poor man's -immutable here- objects) or a global dict state or multiple states.

Such mechanisms may be insufficient for writing more complex extensions, and it's good to see there are plans to improve rules, but for me Typst evokes the youngster fascinated with TeX and the TeXBook that I once was, the idea of having a simple yet powerful and general system that you can tailor to you needs, so my main hope is for it to avoid the pitfalls that ultimately led into extremely complex dependencies. And it's doing quite fine in that department.

**Reply by laurmaedje** [reply]
These are great observations!

I think the differences between a function and a custom type with a show rule primarily come into play when the semantic element they implement shall be used and modified by various users. This includes the standard library elements, but can also include elements in foundational packages. This way, everybody can build upon the same, well-known structure and configure it with set and show instead of hand-rolling their own function, which is then impossible to further modify.

I agree that many cases of more complex selectors in other languages can be handled simply by using local set and show rules in "components" --- be they implemented via functions or in the future, types. So I don't think all of the proposed rule extensions will necessarily be added, but some are definitely useful.

Once things settle down a bit, we definitely need more official guidance on best practices that touch on these things.
